src/AppBundle/Service/OccurrenceSplitterService.php

```diff
@@ -0,0 +1,179 @@
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+use Symfony\Component\PropertyInfo\Extractor\ReflectionExtractor;
+
+class OccurrenceSplitterService
+{
+    private $dateSeparatorTimezone;
+    private $splitHour;
+    private $splitMinute;
+    private $splitSecond;
+
+    private $occurrenceTraitProperties;
+    private $propertyAccessor;
+
+    /**
+     * OccurrenceSplitterService constructor.
+     *
+     * @param string $dateSeparatorTime
+     * @param string $dateSeparatorTimezone
+     */
+    public function __construct(string $dateSeparatorTime, string $dateSeparatorTimezone)
+    {
+        $exploded = explode(':', $dateSeparatorTime);
+        $this->splitHour = (int) $exploded[0];
+        $this->splitMinute = (int) $exploded[1];
+        $this->splitSecond = (int) $exploded[2];
+        $this->dateSeparatorTimezone = new \DateTimeZone($dateSeparatorTimezone);
+
+        $propertyInfo = new ReflectionExtractor();
+        $this->occurrenceTraitProperties = $propertyInfo->getProperties(OccurrenceTrait::class);
+
+        $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
+    }
+
+    /**
+     * Get new DailyOccurrences from an Occurrence.
+     *
+     * @param Occurrence $occurrence
+     *
+     * @throws \Exception
+     *
+     * @return Collection
+     */
+    public function getDailyOccurrences(Occurrence $occurrence): Collection
+    {
+        $dailyOccurrences = new ArrayCollection();
+        if ($occurrence->getStartDate() && $occurrence->getEndDate() && $occurrence->getStartDate() <= $occurrence->getEndDate()) {
+            $this->createDailyOccurrenceCollection($occurrence, $dailyOccurrences);
+        }
+
+        return $dailyOccurrences;
+    }
+
+    /**
+     * Copy values of OccurrenceTrait properties from one DailyOccurrence to another.
+     *
+     * @param DailyOccurrence $to
+     * @param DailyOccurrence $from
+     */
+    public function copyOccurrenceTraitPropertyValues(DailyOccurrence $to, DailyOccurrence $from): void
+    {
+        foreach ($this->occurrenceTraitProperties as $propertyPath) {
+            $value = $this->propertyAccessor->getValue($from, $propertyPath);
+            $this->propertyAccessor->setValue($to, $propertyPath, $value);
+        }
+    }
+
+    /**
+     * Create a collection of DailyOccurrences from an Occurrence.
+     *
+     * @param Occurrence $occurrence
+     * @param Collection $dailyOccurrences
+     *
+     * @throws \Exception
+     *
+     * @return Collection
+     */
+    private function createDailyOccurrenceCollection(Occurrence $occurrence, Collection $dailyOccurrences): Collection
+    {
+        $tempOccurrence = clone $occurrence;
+        $splitDate = $this->getFirstSplitDateTime($tempOccurrence->getStartDate());
+
+        if ($tempOccurrence->getStartDate() < $splitDate && $tempOccurrence->getEndDate() > $splitDate) {
```

Comment:
I think we should add a limit on the number of `DailyOccurrence`s we allow to split into. If an event lasts for 87 years, say, this will run for a very long time.

We could either limit the number of `DailyOccurrence`s created, or skip splitting occurrences that lasts for more than a specific number of days.

Reply:
I agree this is not optimal but disagree with your solution. I think we should either:
* Leave as is
* Limit the allowed length of Occurrences to `< 1 year` or `< 10 years`
* Implement a job queue and handle splitting asynchronously.

I think having an implementation were `DailyOccurrences` match their respective `Occurrence` "except if..." is the wrong way to go on this.